examples/generative/molecule_generation.py

```diff
@@ -0,0 +1,493 @@
+
+SMILE_CHARSET = ast.literal_eval(SMILE_CHARSET)
+SMILE_CHARSET.append("\\")
+
+MAX_MOLSIZE = max(df["smiles"].str.len())
+SMILE_to_index = dict((c, i) for i, c in enumerate(SMILE_CHARSET))
+index_to_SMILE = dict((i, c) for i, c in enumerate(SMILE_CHARSET))
+print("Max molecule size: {}".format(MAX_MOLSIZE))
+print("Character set Length: {}".format(len(SMILE_CHARSET)))
+
+"""
+## Hyperparameters
+"""
+
+latent_dim = 292
+BATCH_SIZE = 256
+EPOCHS = 50
+MAX_LEN = 120
+VAE_LR = 1e-4
+
+"""
+## Build Dataloader
+"""
+
+
+class DataGenerator(tf.keras.utils.Sequence):
+    def __init__(self, data, mapping, max_len, batch_size=6, shuffle=True):
+        """
+        Initialization
+        """
+        self.data = data
+        self.indices = self.data.index.tolist()
+        self.mapping = mapping
+        self.max_len = max_len
+        self.batch_size = batch_size
+        self.shuffle = shuffle
+
+        self.on_epoch_end()
+
+    def __len__(self):
+        return int(np.ceil(len(self.data) / self.batch_size))
+
+    def __getitem__(self, index):
+        if (index + 1) * self.batch_size > len(self.indices):
+            self.batch_size = len(self.indices) - index * self.batch_size
+        # Generate one batch of data
+        # Generate indices of the batch
+        index = self.indices[index * self.batch_size : (index + 1) * self.batch_size]
+        # Find list of IDs
+        batch = [self.indices[k] for k in index]
+        smile, qed = self.data_generation(batch)
+
+        return smile, qed
+
+    def on_epoch_end(self):
+
+        """
+        Updates indexes after each epoch
+        """
+        self.index = np.arange(len(self.indices))
+        if self.shuffle == True:
+            np.random.shuffle(self.index)
+
+    def load(self, idx):
+        """
+        Load molecules in SMILE representation
+        and their respective QED value.
+        """
+        item = self.data.loc[idx]["smiles"]
+        item = list(item)
+        item = [*map(self.mapping.get, item)]
+
+        # padding
+        item = item + [0] * abs((len(item) - self.max_len))
+
+        # One hot Encoding
+        item = tf.one_hot(item, depth=34)
+
+        qed = self.data.loc[idx]["qed"]
+
+        return item, qed
+
+    def data_generation(self, batch):
+
+        x0 = np.empty((self.batch_size, self.max_len, len(self.mapping)))
+        x1 = np.empty((self.batch_size,))
+
+        for i, batch_id in enumerate(batch):
+            (
+                x0[
+                    i,
+                ],
+                x1[
+                    i,
+                ],
+            ) = self.load(batch_id)
+
+        return x0, x1
+
+
+train_df = df.sample(frac=0.75, random_state=42)  # random state is a seed value
+test_df = df.drop(train_df.index)
+train_df.reset_index(drop=True, inplace=True)
+test_df.reset_index(drop=True, inplace=True)
+
+"""
+# Build the Encoder and Decoder
+"""
+
+
+class Encoder(tf.keras.Model):
```

Comment:
The encoder and decoder classes would be simpler and shorter if implemented as Functional models. Right now layers are defined/used twice. See: https://keras.io/examples/generative/vae/

You could package them as functions `get_encoder()` etc.